lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -1343,14 +1372,56 @@ struct LowerTypesPass : public LowerFIRRTLTypesBase<LowerTypesPass> {
 // This is the main entrypoint for the lowering pass.
 void LowerTypesPass::runOnOperation() {
   std::vector<Operation *> ops;
-  llvm::for_each(getOperation().getBody()->getOperations(),
-                 [&](Operation &op) { ops.push_back(&op); });
+  // Record of each module, to the map of NonLocalAnchors which need to be
+  // updated with the lowered inner_sym name.
+  DenseMap<Operation *, DenseMap<StringAttr, StringAttr>>
+      modToNlaNameToNewSymMap;
+  // Map of name of the NonLocalAnchor to the operation.
+  DenseMap<StringAttr, Operation *> nlaMap;
+  // Record all operations in the circuit.
+  llvm::for_each(getOperation().getBody()->getOperations(), [&](Operation &op) {
+    ops.push_back(&op);
+    // Creating a map of all ops in the circt, but only modules are relevant.
+    modToNlaNameToNewSymMap.insert(
+        std::make_pair<Operation *, DenseMap<StringAttr, StringAttr>>(
+            &op, /*An empty map*/ DenseMap<StringAttr, StringAttr>()));
+    // Record the NonLocalAnchor and its name.
+    if (auto nla = dyn_cast<NonLocalAnchor>(op))
+      nlaMap[nla.sym_nameAttr()] = nla;
+  });
 
+  // Each parallel module worker updates the map to the corresponding module.
+  // After this statement runs, modToNlaNameToNewSymMap[ops[index]] will be
+  // updated with all NonLocalAnchors that need to be updated with the lowerd
+  // symbols.
   mlir::parallelForEachN(&getContext(), 0, ops.size(), [&](auto index) {
     TypeLoweringVisitor(&getContext(), flattenAggregateMemData,
-                        preserveAggregate, preservePublicTypes)
+                        preserveAggregate, preservePublicTypes,
+                        modToNlaNameToNewSymMap[ops[index]])
         .lowerModule(ops[index]);
   });
+  // Fixup the nla, with the updated symbol names.
+  // This can only update the final element on which the nla is applied, because
+  // lowering cannot update the symbols on the InstanceOp. Also, the final
+  // element cannot be a module.
+  for (auto modMap : modToNlaNameToNewSymMap) {
+    for (auto nlaToSym : modMap.second) {
+      // Get the nla with the corresponding name. This is guaranteed to exist.
+      // nlaToSym.first is the NLA name.
+      auto nla = cast<NonLocalAnchor>(nlaMap[nlaToSym.first]);
+      auto namepath = nla.namepath();
+      SmallVector<Attribute> updatedPath;
+      // Copy the namepath
+      for (auto n : namepath.getValue())
+        updatedPath.push_back(n);
```

Comment:
`SmallVector<Attribute> updatedPath(namepath)` or `(namepath.begin(), namepath.end())` may work. and saves the manual iteration.

Reply:
Done